### ISSUE TITLE
Line Number Accuracy

### ISSUE THREAD

**Issue by JaredCrean2** [JuliaLang/julia#18273]
Line numbers are off by a few lines, causing confusion.  Here is a minimal standalone example:

``` julia
module C

  global const PETSC_DECIDE = -1
  typealias PetscInt Int64
  MATMPIAIJ = "matmpiaij"
  MatType = Symbol


end  # end module C

module MPI
  typealias Comm Cint
  COMM_WORLD = Cint(42)
end

module PETSc

using MPI
using C

function Mat{T}(::Type{T}, mtype::C.MatType=C.MATSEQAIJ; comm::MPI.Comm=MPI.COMM_WORLD)

  println("I found the other half of the problem")
end


function Mat{T}(::Type{T}, m::Integer=C.PETSC_DECIDE, n::Integer=C.PETSC_DECIDE;
                mlocal::Integer=C.PETSC_DECIDE, nlocal::Integer=C.PETSC_DECIDE,
                bs=1, nz::Integer=0, nnz::AbstractVector=PetscInt[],
                onz::Integer=0, onnz::AbstractVector=PetscInt[],
                comm::MPI.Comm=MPI.COMM_WORLD,
                mtype::Symbol=C.MATMPIAIJ)


  println("I finally found the problem")

  return nothing
end

end  # end module PETSc

using PETSc

```

gives output:

```
WARNING: Method definition Mat(Type{#T<:Any}) in module PETSc at /home/jared/.julia/v0.4/PETSc_current/src/tmp.jl:24 overwritten at /home/jared/.julia/v0.4/PETSc_current/src/tmp.jl:36.
WARNING: Method definition #Mat(Array{Any, 1}, PETSc.#Mat, Type{#T<:Any}) in module PETSc overwritten.
```

lines 24 and 36 contain the `println` statements, rather than the function declarations.

**Comment by JaredCrean2** [comment]
I should add this occurred with both 0.5-rc1 and the more recent

```julia
Julia Version 0.6.0-dev.201
Commit e4b5233* (2016-08-13 16:10 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7 CPU       Q 820  @ 1.73GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, nehalem)
```

**Comment by yuyichao** [comment]
Note that fixing this should be careful to not break coverage line number.

**Comment by JeffBezanson** [comment]
This has come up before --- we only keep the line number for the first line of code in the function, since that's the one we need for coverage and error reporting. It still uniquely and correctly identifies the function so I don't consider this a significant problem.

**Comment by JaredCrean2** [comment]
The line number of the first statement does uniquely identify the function, but incorrectly identifies the error. For example, in the original code the second function was calling the first function on the first line, so it appeared calling the function was overwriting it.

**Comment by yuyichao** [comment]
The warning in this case isn't really confusing since it's pretty clear where the override is. It could be more confusing if the first line is a function definition though.

> It still uniquely and correctly identifies the function

Not actually true when there's nested function definition. It is possible to define multiple functions in one line but the currently lineno handling does cause some false positive.

**Comment by evanfields** [comment]
Line numbers are sometimes off by more than a few lines. Right now I'm working on a script (180 lines) but when I include the script, I get a method error referencing an expression starting on line 480. If I can create a small example I'll update this comment.

```
julia> include("src\\simulation_based_optimization\\so.jl")
ERROR: LoadError: UndefVarError: SOPoinType not defined
 in include_from_node1(::String) at .\loading.jl:426
while loading C:\Users\ejfie\zipcarpodmodel\src\simulation_based_optimization\so.jl, in expression starting on line 480
```

Maybe that 480 refers to something in `loading.jl` in which case it's an issue of deceptive error messages, not line accuracy.

**Comment by yuyichao** [comment]
@evanfields this is a different issue. I mostly see this happen when there's macros that messes up line number info. Please also include the exact version of julia you are using when you report the issue.

**Comment by vtjnash** [comment]
Confirmed fixed by #35138
